yapapi/props/builder.py

```diff
@@ -87,3 +79,21 @@ def add(self, m: Model):
     async def subscribe(self, market: Market) -> Subscription:
         """Create a Demand on the market and subscribe to Offers that will match that Demand."""
         return await market.subscribe(self._properties, self.constraints)
+
+    async def decorate(self, *decorators: 'DemandDecorator'):
+        for decorator in decorators:
+            await decorator.decorate_demand(self)
+
+
+class DemandDecorator(abc.ABC):
+    """An interface that specifies classes that can add properties and constraints through a DemandBuilder"""
+
+    @abc.abstractmethod
+    async def decorate_demand(self, demand: DemandBuilder):
+        """Add appropriate properties and constraints to a Demand"""
+
+
+class AutodecoratingModel(Model, DemandDecorator):
```

Comment:
Can't we make any `Model` "autodecorating"?

Reply:
hmm... I dunno... I thought about it and wasn't sure if it really makes sense a) for each Model to be a potential DemandDecorator - and a b) one that does that based on its own properties ... maybe it's better to make it more explicit....